File: android/src/newarch/com/zoontek/rnpermissions/RNPermissionsModule.java
package com.zoontek.rnpermissions;

import android.util.SparseArray;

import androidx.annotation.NonNull;

import com.facebook.react.bridge.Callback;
import com.facebook.react.bridge.Promise;
import com.facebook.react.bridge.ReactApplicationContext;
import com.facebook.react.bridge.ReadableArray;
import com.facebook.react.module.annotations.ReactModule;
import com.facebook.react.modules.core.PermissionListener;

import java.util.Map;

@ReactModule(name = RNPermissionsModuleImpl.NAME)
public class RNPermissionsModule extends NativeRNPermissionsSpec implements PermissionListener {

  private final SparseArray<Callback> mCallbacks;

  public RNPermissionsModule(ReactApplicationContext reactContext) {
    super(reactContext);
    mCallbacks = new SparseArray<Callback>();
  }

  @Override
  @NonNull
  public String getName() {
    return RNPermissionsModuleImpl.NAME;
  }

  @Override
  protected Map<String, Object> getTypedExportedConstants() {
    return RNPermissionsModuleImpl.getConstants();
  }

  @Override
  public void openSettings(Promise promise) {
    RNPermissionsModuleImpl.openSettings(getReactApplicationContext(), promise);
  }

  @Override
  public void check(String permission, Promise promise) {
    RNPermissionsModuleImpl.check(getReactApplicationContext(), permission, promise);
  }

  @Override
  public void checkNotifications(Promise promise) {
    RNPermissionsModuleImpl.checkNotifications(getReactApplicationContext(), promise);
  }

  @Override
  public void checkMultiple(ReadableArray permissions, Promise promise) {
    RNPermissionsModuleImpl.checkMultiple(getReactApplicationContext(), permissions, promise);
  }

  @Override
  public void request(String permission, Promise promise) {
    RNPermissionsModuleImpl.request(getReactApplicationContext(), this, mCallbacks, permission, promise);
  }

  @Override
  public void requestNotifications(ReadableArray options, Promise promise) {
    RNPermissionsModuleImpl.requestNotifications(getReactApplicationContext(), promise);
  }

  @Override
  public void requestMultiple(ReadableArray permissions, Promise promise) {
    RNPermissionsModuleImpl.requestMultiple(getReactApplicationContext(), this, mCallbacks, permissions, promise);
  }

  @Override
  public void shouldShowRequestRationale(String permission, Promise promise) {
    RNPermissionsModuleImpl.shouldShowRequestRationale(getReactApplicationContext(), permission, promise);
  }

  @Override
  public void checkLocationAccuracy(Promise promise) {
    RNPermissionsModuleImpl.checkLocationAccuracy(promise);
  }

  @Override
  public void requestLocationAccuracy(String purposeKey, Promise promise) {
    RNPermissionsModuleImpl.requestLocationAccuracy(promise);
  }

  @Override
  public void openPhotoPicker(Promise promise) {
    RNPermissionsModuleImpl.openPhotoPicker(promise);
  }

  @Override
  public boolean onRequestPermissionsResult(int requestCode, String[] permissions, int[] grantResults) {
    return RNPermissionsModuleImpl.onRequestPermissionsResult(getReactApplicationContext(), mCallbacks, requestCode, grantResults);
  }
}

File: android/src/newarch/com/zoontek/rnpermissions/RNPermissionsModule.kt
package com.zoontek.rnpermissions

import android.util.SparseArray

import com.facebook.react.bridge.Callback
import com.facebook.react.bridge.Promise
import com.facebook.react.bridge.ReactApplicationContext
import com.facebook.react.bridge.ReadableArray
import com.facebook.react.module.annotations.ReactModule
import com.facebook.react.modules.core.PermissionListener

@ReactModule(name = RNPermissionsModuleImpl.NAME)
class RNPermissionsModule(reactContext: ReactApplicationContext?) :
  NativeRNPermissionsSpec(reactContext), PermissionListener {

  private val callbacks = SparseArray<Callback>()

  override fun getName(): String {
    return RNPermissionsModuleImpl.NAME
  }

  override fun openSettings(type: String?, promise: Promise) {
    RNPermissionsModuleImpl.openSettings(reactApplicationContext, type, promise)
  }

  override fun canScheduleExactAlarms(promise: Promise) {
    RNPermissionsModuleImpl.canScheduleExactAlarms(reactApplicationContext, promise)
  }

  override fun check(permission: String, promise: Promise) {
    RNPermissionsModuleImpl.check(reactApplicationContext, permission, promise)
  }

  override fun checkNotifications(promise: Promise) {
    RNPermissionsModuleImpl.checkNotifications(reactApplicationContext, promise)
  }

  override fun checkMultiple(permissions: ReadableArray, promise: Promise) {
    RNPermissionsModuleImpl.checkMultiple(reactApplicationContext, permissions, promise)
  }

  override fun request(permission: String, promise: Promise) {
    RNPermissionsModuleImpl.request(reactApplicationContext, this, callbacks, permission, promise)
  }

  override fun requestNotifications(options: ReadableArray, promise: Promise) {
    RNPermissionsModuleImpl.requestNotifications(reactApplicationContext, promise)
  }

  override fun requestMultiple(permissions: ReadableArray, promise: Promise) {
    RNPermissionsModuleImpl.requestMultiple(reactApplicationContext, this, callbacks, permissions, promise)
  }

  override fun shouldShowRequestRationale(permission: String, promise: Promise) {
    RNPermissionsModuleImpl.shouldShowRequestRationale(reactApplicationContext, permission, promise)
  }

  override fun checkLocationAccuracy(promise: Promise) {
    RNPermissionsModuleImpl.checkLocationAccuracy(promise)
  }

  override fun requestLocationAccuracy(purposeKey: String, promise: Promise) {
    RNPermissionsModuleImpl.requestLocationAccuracy(promise)
  }

  override fun openPhotoPicker(promise: Promise) {
    RNPermissionsModuleImpl.openPhotoPicker(promise)
  }

  override fun onRequestPermissionsResult(requestCode: Int, permissions: Array<String>, grantResults: IntArray): Boolean {
    return RNPermissionsModuleImpl.onRequestPermissionsResult(reactApplicationContext, callbacks, requestCode, grantResults)
  }
}
